s4e-web/src/app/views/settings/breadcrumb/breadcrumb.service.ts

```diff
@@ -0,0 +1,105 @@
+import { map, switchMap } from 'rxjs/operators';
+import { Observable, of, BehaviorSubject } from 'rxjs';
+import { IBreadcrumb } from './breadcrumb.model';
+import { Routes, Route, ActivatedRoute, Router } from '@angular/router';
+import { Injectable } from '@angular/core';
+
+@Injectable({
+  providedIn: 'root'
+})
+export class BreadcrumbService {
+  static _MAX_DEPTH = 8;
+
+  public defaultRoute: Route;
+  public rootPath = '';
+
+  public breadcrumbs$: Observable<IBreadcrumb[]>;
+  protected _breadcrumbSubject$: BehaviorSubject<IBreadcrumb[]> = new BehaviorSubject<IBreadcrumb[]>([]);
+
+  constructor() {
+    this.breadcrumbs$ = this._breadcrumbSubject$.asObservable();
+  }
+
+  registerRoutes(childrenRoutes: Routes, rootPath = '') {
+    this.defaultRoute = this._getDefaultRouteFrom(childrenRoutes);
+    this.rootPath = rootPath;
+  }
+  loadFrom(activatedRoute: ActivatedRoute) {
+    return this._getBreadcrumbsFrom(this._getActualRouteFrom(activatedRoute))
+      .subscribe((breadcrumbs: IBreadcrumb[]) => this._breadcrumbSubject$.next(breadcrumbs));
+  }
+
+  _getBreadcrumbsFrom(childRoute: ActivatedRoute, breadcrumbs: IBreadcrumb[] = [], depth = 0): Observable<IBreadcrumb[]> {
```

Comment:
If this is private then mark it as `private` (or `protected`), same for others

Reply:
The problem occurs while testing. I can't access public methods and testing the main function, because it is too complex and recursive. On other hand, I want to say other developers that those methods are created only for internal usage. If you have any proposal in such case, I'm open on it...